File: app/src/main/java/com/devmatic/fruits/data/api/ApiUtils.java
package com.devmatic.fruits.data.api;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.util.*;
public class ApiUtils {

    public static <T> T fromJson(Object response, Class cls) {
        return fromJson(response, cls, new Gson());
    }

    public static <T> T fromJson(Object response, Class cls, Gson gson) {
        String name = response.getClass().getSimpleName();
        return (T) gson.fromJson(String.valueOf(response), cls);
    }

    public static String toJson(Object object) {
        return toJson(object, new Gson());
    }

    public static String toJson(Object object, Gson gson) {
        return gson.toJson(object);
    }

    public static <T> ArrayList<T> convertArray(JsonArray jArr, Class<?> clazz) {
        ArrayList<T> list = new ArrayList<T>();
        try {
            for (int i = 0, l = jArr.size(); i < l; i++) {
                list.add((T) new Gson().fromJson(jArr.get(i), clazz));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return list;
    }

    public static HashMap<String, String> getJsonObjectToHashMap(JSONObject params) {
        HashMap<String, String> mapParams = new HashMap<>();
        try {
            for (Map.Entry<String, Object> entry : jsonToMap(params).entrySet()) {
                mapParams.put(entry.getKey(), String.valueOf(entry.getValue()));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return mapParams;
    }


    public static Map<String, Object> jsonToMap(JSONObject json) throws JSONException {
        Map<String, Object> retMap = new HashMap<>();
        if(json != JSONObject.NULL) {
            retMap = toMap(json);
        }
        return retMap;
    }

    public static Map<String, Object> toMap(JSONObject object) throws JSONException {
        Map<String, Object> map = new HashMap<>();
        Iterator<String> keysItr = object.keys();
        while(keysItr.hasNext()) {
            String key = keysItr.next();
            Object value = object.get(key);
            if(value instanceof JSONArray) {
                value = toList((JSONArray) value);
            }
            else if(value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            map.put(key, value);
        }
        return map;
    }

    public static List<Object> toList(JSONArray array) throws JSONException {
        List<Object> list = new ArrayList<>();
        for(int i = 0; i < array.length(); i++) {
            Object value = array.get(i);
            if(value instanceof JSONArray) {
                value = toList((JSONArray) value);
            }
            else if(value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            list.add(value);
        }
        return list;
    }


    public static boolean isHostAvailable(String host, int port, int timeoutMs) {
        try {
            Socket sock = new Socket();
            SocketAddress sockaddr = new InetSocketAddress(host, port);
            sock.connect(sockaddr, timeoutMs); // This will block no more than timeoutMs
            sock.close();
            return true;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    public static void sendSocketData(String host, int port, String data){
        try {
            // The line below illustrates the default port 6101 for mobile printers 9100 is the default port number
            // for desktop and tabletop printers
            Socket clientSocket=new Socket(host, port);
            DataOutputStream dos = new DataOutputStream(clientSocket.getOutputStream() );
            //The data being sent in the lines below illustrate CPCL  one can change the data for the corresponding
            //language being used (ZPL, EPL)
            dos.writeBytes(data);
//            dos.writeUTF("If procrastination was a sport, I would compete in it later.");
            clientSocket.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/devmatic/fruits/data/api/ApiUtils.kt
package com.devmatic.fruits.data.api

import com.google.gson.Gson
import com.google.gson.JsonArray
import okhttp3.MediaType
import okhttp3.RequestBody
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import retrofit2.HttpException
import java.util.*


fun <T> fromJson(response: Any, cls: Class<*>): T {
    return fromJson(response, cls, Gson())
}

fun <T> fromJson(response: Any, cls: Class<*>, gson: Gson): T {
    val name = response.javaClass.simpleName
    val t = gson.fromJson(response.toString(), cls) as T
    return t
}

@JvmOverloads
fun toJson(obj: Any, gson: Gson = Gson()): String {
    return gson.toJson(obj)
}

fun jsonObjectFillHashMap(vararg params: JSONObject?, map: HashMap<String, Any>): HashMap<String, Any> {
    for (param in params) {
        if (param != null) {
            map.putAll(getJsonObjectToHashMap(param))
        }
    }
    return map
}

fun <T> convertArray(jArr: JsonArray, clazz: Class<Any>): ArrayList<T> {
    val list = ArrayList<T>()
    try {
        var i = 0
        val l = jArr.size()
        while (i < l) {
            list.add(Gson().fromJson<Any>(jArr.get(i), clazz) as T)
            i++
        }
    } catch (e: Exception) {
        e.printStackTrace()
    }

    return list
}

fun getRequestBody(params: HashMap<String, Any>): RequestBody {
    return RequestBody.create(MediaType.parse("application/json; charset=utf-8"), JSONObject(params).toString())
}


fun getJsonObjectToHashMap(params: JSONObject): HashMap<String, Any> {
    val mapParams = HashMap<String, Any>()
    try {
        for ((key, value) in jsonToMap(params)) {
            mapParams[key] = value
        }
    } catch (e: Exception) {
        e.printStackTrace()
    }
    return mapParams
}

@Throws(JSONException::class)
fun jsonToMap(json: JSONObject): Map<String, Any> {
    var retMap: Map<String, Any> = HashMap()
    if (json !== JSONObject.NULL) {
        retMap = toMap(json)
    }
    return retMap
}

@Throws(JSONException::class)
private fun toMap(obj: JSONObject): Map<String, Any> {
    val map = HashMap<String, Any>()
    val keysItr = obj.keys()
    while (keysItr.hasNext()) {
        val key = keysItr.next()
        var value = obj.get(key)
        if (value is JSONArray) {
            value = toList(value)
        } else if (value is JSONObject) {
            value = toMap(value)
        }
        map[key] = value
    }
    return map
}

@Throws(JSONException::class)
private fun toList(array: JSONArray): List<Any> {
    val list = ArrayList<Any>()
    for (i in 0 until array.length()) {
        var value = array.get(i)
        if (value is JSONArray) {
            value = toList(value)
        } else if (value is JSONObject) {
            value = toMap(value)
        }
        list.add(value)
    }
    return list
}


fun getResponseError(throwable: Throwable): String {
    var error = ""
    var code: Int
    try {
        if (throwable is HttpException) {
            code = throwable.code()
            val result = throwable.response()?.errorBody()?.string()
            try {
                val responseError = JSONObject(result)
                code = responseError.getInt("code")
                val message = StringBuilder()
                message.append("$code;")
                if (responseError.has("error")) {
                    message.append(responseError.getString("error")).append(";")
                }
                if (responseError.has("message")) {
                    message.append("Message:${responseError.getString("message").trim()}").append(";")
                }
                if (throwable.message != null) {
                    message.append("Ошибка сервера:${throwable.message}").append(";")
                }
                error = message.toString()
            } catch (e: JSONException) {
                e.printStackTrace()
                error = result.toString()
            }

            when (code) {
                504 -> error = "Время ожидания истекло,повторите запрос позже"
                503 -> error = "Сервис временно недоступен,повторите запрос позже"
            }
        } else {
            val message = throwable.message ?: "Ошибка запроса"
            val contains = when {
                message.contains("Unable to resolve host") -> true
                message.contains("Failed to connect") -> true
                else -> false
            }
            error = if (contains) {
                "Ошибка соединения, адрес недоступен"
            } else {
                message
            }
        }
    } catch (e: Exception) {
        e.printStackTrace()
    }
    return error
}